File: src/main/java/com/nanaios/polygonal_tech/capability/impl/LongFluidTank.java
package com.nanaios.polygonal_tech.capability.impl;

import com.nanaios.polygonal_tech.capability.interfaces.ILongFluidTank;
import com.nanaios.polygonal_tech.fluids.base.LongFluidStack;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.fluids.capability.IFluidHandler.FluidAction;
import org.jetbrains.annotations.NotNull;

import java.util.function.LongSupplier;
import java.util.function.Predicate;

public class LongFluidTank implements ILongFluidTank {
    protected final LongSupplier capacity;
    protected final Predicate<FluidStack> validator;
    protected LongFluidStack fluid = LongFluidStack.EMPTY;

    public LongFluidTank(LongSupplier capacity, Predicate<FluidStack> validator) {
        this.capacity = capacity;
        this.validator = validator;
    }

    @Override
    public long getFluidLongAmount() {
        return fluid.getAmount();
    }

    @Override
    public long getLongCapacity() {
        return capacity.getAsLong();
    }

    @Override
    public long fillLong(LongFluidStack resource, FluidAction action) {
        if (resource.isEmpty() || !isFluidValid(resource)) return 0;

        if (action.simulate()) {
            if (fluid.isEmpty()) return Math.min(capacity.getAsLong(), resource.getAmount());
            if (!fluid.isFluidEqual(resource)) return 0;
            return Math.min(capacity.getAsLong() - fluid.getAmount(), resource.getAmount());
        }

        if (fluid.isEmpty()) {
            fluid = new LongFluidStack(resource, Math.min(capacity.getAsLong(), resource.getAmount()));
            onContentsChanged();
            return fluid.getLongAmount();
        }

        if (!fluid.isFluidEqual(resource)) return 0;

        long filled = capacity.getAsLong() - fluid.getLongAmount();

        if (resource.getAmount() < filled) {
            fluid.grow(resource.getLongAmount());
            filled = resource.getLongAmount();
        } else {
            fluid.setAmount(capacity.getAsLong());
        }

        if (filled > 0) onContentsChanged();
        return filled;
    }

    public @NotNull LongFluidStack drain(LongFluidStack resource, FluidAction action) {
        if (resource.isEmpty() || !resource.isFluidEqual(getFluid())) return LongFluidStack.EMPTY;
        return drain(resource.getLongAmount(), action);
    }


    @Override
    public @NotNull LongFluidStack getFluid() {
        return fluid;
    }

    @Override
    public boolean isFluidValid(FluidStack stack) {
        return validator.test(stack);
    }

    @Override
    public LongFluidStack drain(long maxDrain, FluidAction action) {

        long drained = maxDrain;
        if (fluid.getLongAmount() < drained) {
            drained = fluid.getLongAmount();
        }

        LongFluidStack stack = new LongFluidStack(fluid, drained);
        if (action.execute() && drained > 0) {
            fluid.shrink(drained);
            onContentsChanged();
        }

        return stack;
    }

    protected void onContentsChanged() {

    }

    public static class InputOnly extends LongFluidTank {
        protected final ILongFluidTank base;
        public InputOnly(LongFluidTank base) {
            super(base.capacity, base.validator);
            this.base = base;
        }

        @Override
        public long getFluidLongAmount() {
            return base.getFluidLongAmount();
        }

        @Override
        public @NotNull LongFluidStack getFluid() {
            return base.getFluid();
        }

        @Override
        public LongFluidStack drain(long maxDrain, FluidAction action) {
            return LongFluidStack.EMPTY;
        }

        @Override
        public long fillLong(LongFluidStack resource, FluidAction action) {
            return base.fillLong(resource, action);
        }

        @Override
        public boolean isFluidValid(FluidStack stack) {
            return base.isFluidValid(stack);
        }
    }

    public static class OutputOnly extends LongFluidTank {
        protected final ILongFluidTank base;
        public OutputOnly(LongFluidTank base) {
            super(base.capacity, base.validator);
            this.base = base;
        }

        @Override
        public long getFluidLongAmount() {
            return base.getFluidLongAmount();
        }

        @Override
        public @NotNull LongFluidStack getFluid() {
            return base.getFluid();
        }

        @Override
        public LongFluidStack drain(long maxDrain, FluidAction action) {
            return base.drain(maxDrain, action);
        }

        @Override
        public long fillLong(LongFluidStack resource, FluidAction action) {
            return 0;
        }

        @Override
        public boolean isFluidValid(FluidStack stack) {
            return false;
        }
    }
}

File: src/main/kotlin/com/nanaios/polygonal_tech/core/capability/fluid/LongFluidTank.kt
package com.nanaios.polygonal_tech.core.capability.fluid

import com.nanaios.polygonal_tech.core.fluid.LongFluidStack
import com.nanaios.polygonal_tech.core.fluid.readLongFluidStack
import com.nanaios.polygonal_tech.core.fluid.writeLongFluidStack
import com.nanaios.polygonal_tech.core.network.sync.storage.EmptySyncValueStorage
import com.nanaios.polygonal_tech.core.network.sync.storage.ISyncValueStorage
import com.nanaios.polygonal_tech.core.network.sync.type.ISyncType
import com.nanaios.polygonal_tech.core.network.sync.type.SyncType
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.FriendlyByteBuf
import net.minecraftforge.fluids.capability.IFluidHandler
import kotlin.math.min

/**
 * [ILongFluidTank]の標準的な実装クラスであり、
 * NBTへのセーブ/ロード、ネットワーク同期（[ISyncValue]としての振る舞い）、大容量流体のシミュレート及び実更新ロジックを統合提供することを目的とする。
 *
 * @param fluid 初期の流体内容
 * @param amount 初期の流体量
 * @param capacity このタンクの最大容量
 * @param valid 注入されようとしている流体が条件に合うか（熱量のみ、水のみ等）を検証する述語関数
 */
open class LongFluidTank(
    fluid: LongFluidStack,
    amount: Long,
    capacity: Long,
    protected val valid: (LongFluidStack) -> Boolean
) : ILongFluidTank {
    override var storage: ISyncValueStorage = EmptySyncValueStorage
    override var type: ISyncType = SyncType.NONE
    override var isDirty = false
        protected set
    override var longFluid = fluid
        protected set
    override val longFluidAmount
        get() = longFluid.getAmount()
    override var longCapability = capacity
        protected set

    protected fun markDirty() {
        isDirty = true
        storage.onSyncValueChanged(this)
    }

    override fun serializeNBT(): CompoundTag {
        val tag = CompoundTag()
        longFluid.writeToNBT(tag)
        return tag
    }

    override fun deserializeNBT(nbt: CompoundTag) {
        longFluid = LongFluidStack.loadFluidStackFromNBT(nbt)
    }

    override fun writeBuffer(buffer: FriendlyByteBuf) {
        buffer.writeLongFluidStack(longFluid)
        buffer.writeLong(longCapability)
    }

    override fun readBuffer(buffer: FriendlyByteBuf) {
        longFluid = buffer.readLongFluidStack()
        longCapability = buffer.readLong()
    }

    override fun onSync() {
        isDirty = false
    }

    override fun isFluidValid(stack: LongFluidStack): Boolean {
        return valid(stack)
    }

    override fun fill(
        resource: LongFluidStack,
        action: IFluidHandler.FluidAction
    ): Long {
        if (resource.isEmpty() || !isFluidValid(resource)) {
            return 0
        }
        if (action.simulate()) {
            if (longFluid.isEmpty()) {
                return min(longCapability, resource.getAmount())
            }
            if (!longFluid.isFluidEqual(resource)) {
                return 0
            }
            return min(longCapability - longFluid.getAmount(), resource.getAmount())
        }
        if (longFluid.isEmpty()) {
            longFluid = LongFluidStack(resource.fluid, min(longCapability, resource.getAmount()))
            markDirty()
            return longFluid.getAmount()
        }
        if (!longFluid.isFluidEqual(resource)) {
            return 0
        }
        var filled = longCapability - longFluid.getAmount()

        if (resource.getAmount() < filled) {
            longFluid.grow(resource.getAmount())
            filled = resource.getAmount()
        } else {
            fluid.setAmount(capacity)
        }
        if (filled > 0) markDirty()
        return filled
    }

    override fun drain(
        maxDrain: Long,
        action: IFluidHandler.FluidAction
    ): LongFluidStack {
        var drained = maxDrain
        if (longFluid.getAmount() < drained) {
            drained = longFluid.getAmount()
        }
        val stack = LongFluidStack(longFluid.fluid, drained)
        if (action.execute() && drained > 0) {
            longFluid.shrink(drained)
            markDirty()
        }
        return stack
    }

    override fun drain(
        resource: LongFluidStack,
        action: IFluidHandler.FluidAction
    ): LongFluidStack {
        if (resource.isEmpty() || !resource.isFluidEqual(fluid)) {
            return LongFluidStack.EMPTY
        }
        return drain(resource.getAmount(), action)
    }
}